### ISSUE TITLE
Autoload global variables behaving inconsistently, properties not having types inferred

### ISSUE THREAD

**Issue by nazjun** [godotengine/godot#77643]
### Godot version

v4.0.3.stable.mono.official [5222a99f5]

### System information

- OS: Windows 10
- Godot Version: 4.0.3.stable.mono
- Rendering Driver: Any
- Rendering Method: Any

### Issue description

I've noticed some strange behaviors arising from the use of static typing (along with type casting in general) with autoload global variables.

Here are my findings so far:
- Creating an autoload from a .tscn file will cause the singleton node referenced by the global variable to be unconvertible to the type of the script specified after `class_name`, if it exists.
- Creating an autoload from a .gd file, however, _does_ allow the variable to be converted to the custom type.
- Properties of custom reference types in an autoload will _not_ be inferred with the `:=` operator if the autoload is .tscn-based, but _will_ if it is .gd-based.
- Properties of custom reference types in an autoload class will not have their own custom properties appear in the auto-completion menu, if the autoload is accessed with its global variable. The auto-completion menu will only reflect the built-in classes they're derived from. Creating a variable of the singleton's custom type (which is only possible with .gd autoloads) and then accessing the reference property that way is the only way I can find that allows the auto-completion to accurately reflect the current type of the reference.

The best way I can work around this while maintaining type safety is by only using .gd-based autoloads, but unless I store an explicit reference to the singleton as its proper type, I won't be able to see the full classes of its custom properties.

### Steps to reproduce

In the main scene's script, `main.gd`:
```gdscript
class_name Main extends Node

var number: int = 7

func _enter_tree() -> void:
    Global.main = self
    Global2.main = self

func _ready() -> void:
    # Does not autofill number property, will also fail to infer type
    var num_0 := Global.main.number
    
    # Autofills number property, but casting fails when using a .tscn autoload
    var global_1 := Global as _Global
    var num_1 := global_1.main.number
    
    # Autofills number property, and casting succeeds when using a .gd autoload
    var global_2 := Global2 as _Global2
    var num_2 := global_2.main.number
```

In the first singleton script, `global.gd`, loaded as a **.tscn**:
```gdscript
class_name _Global extends Node

var main: Main
```

In the first singleton script, `global_2.gd`, loaded as a **.gd**:
```gdscript
class_name _Global2 extends Node

var main: Main
```

The autoload configuration:
![image](https://github.com/godotengine/godot/assets/31663251/749add78-46fc-439e-acf3-da8e2e126edb)

### Minimal reproduction project

[TypeBug.zip](https://github.com/godotengine/godot/files/11596664/TypeBug.zip)

**Comment by nazjun** [comment]
> > Creating an autoload from a .tscn file will cause the singleton node referenced by the global variable to be unconvertible to the type of the script specified after class_name, if it exists.
> 
> What error if any is thrown? Does this match or not the behaviour when having a node with the same script attached that is not autoload

That line of code throws the `Parser Error: Invalid cast. Cannot convert from "" to "_Global".` error.

Creating a child node called `NotGlobal`, then adding the following code does not throw any errors, but it _also_ does not autocomplete the `main` property that is part of the `_Global` class, despite the static declaration:
```gdscript
@onready var not_global: _Global = $NotGlobal

func _ready() -> void:
    var num_e = not_global.main.number
```
This matches the erroneous behavior described in #73638, where assigning the result of `get_node` (or its shorthand) to a statically typed variable will override the variable with the base type, instead of the one specified in the hint.

**Comment by HolonProduction** [comment]
Related to #78454

**Comment by tracefree** [comment]
I've discovered the same problem, and some additional info described in #98270. Type inference does work as expected with .tscn based autoloads from different scripts, _unless_ those scripts are referenced in the autoloaded script itself. I guess because they are parsed before the autoloaded script has been fully parsed itself then. (Not completely sure how GDScript works internally though.)

**Comment by tracefree** [comment]
I believe I found the root cause of this problem. This line in the analyzer evaluates to `false` when coming across the identifier of the global autoload name that the script currently being analyzed is attached to: https://github.com/godotengine/godot/blob/4631a617e5ab4ec5cc51dbc43609b269a43059d9/modules/gdscript/gdscript_analyzer.cpp#L4461C18-L4461C25

Any suggestion on how to fix this?

**Comment by tracefree** [comment]
Here's the full chain of events leading to the problem in full detail:

1. The engine starts
2. In ``EditorAutoloadSettings::EditorAutoloadSettings()``, the names of autoload singletons get registered, but with a value of `null` instead of instantiating the objects yet
3. In ``EditorAutoloadSettings::init_autoloads()``, for each of the ``AutoloadInfo`` objects the corresponding node is instantiated (using `_create_autoload`). Afterwards these nodes replace the default null value of the autoload singleton names that were created in step 1.
4. However, during instantiation and _before_ assigning the node to the singleton names, the scripts attached to the scene (and scripts that those depend on) are loaded.
5. Those scripts may contain identifiers of the global singleton name currently being instantiated (in my MRP I named it `Autoload` for example).
6. The identifier's datatype is being analyzed in `GDScriptAnalyzer::reduce_identifier`
7. It checks if the identifier refers to an autoload singleton. This check passes, the name was registered in step 2.
8. The corresponding node is retrieved in order to get its script and parse it (with `parser->get_depended_parser_for`).
9. **Here occurs the crucial problem**: The node retrieved is still the default value of `null` assigned in step 2. Further checks therefore fail and the only type information that can be analyzed is that the identifier refers to a node.
10. Type checks on member variables of the autoload therefore fail.

In summary, the analyzer depends on retrieving the node of an autoload to check the path of its script. But the node is not actually assigned to the autoload name yet because we are in the middle of instantiating it for the purpose of assigning it. A classic circular dependency.

**Comment by Kylemcarthur** [comment]
I encountered the same type inference issue described here, but not the autocomplete issues. I also found it was even affecting things like the autoload's consts erroring as "not constant expressions" at runtime when referenced elsewhere. However, this was inconsistent, as the reference was working fine for many days, then suddenly starting throwing that error today without any other changes.

Following the findings in this thread, until a fix for this gets implemented, the issue does seem to resolve if you are able to switch the autoload in question to a .gd instead of a .tscn. If your autoload is depending on it being a scene, this won't help you. Switching the error-causing autoload to a .gd resolved both the type inference and const problems described.

**Comment by hsandt** [comment]
Looks related: https://github.com/godotengine/godot/issues/114422

**Comment by tracefree** [comment]
> Looks related: [#114422](https://github.com/godotengine/godot/issues/114422)

Possibly. I tested it with my PR and it does not fix the issue with the MRP posted there unfortunately.
